### PR TITLE
mnaveau/solo12

### PR DESCRIPTION
**Pull request by MaximilienNaveau**
## Description

This PR deals with the newcomers Solo12.

## Changes done in the PR

- Rename solo in solo8 everywhere
- Introduce all solo12 equivalent files.

## How I Tested

- run the demo on the solo12 robot: sudo ./demo_solo12 network_id
- run the demo bounding using the DynamicGraphManager on solo8.

## Do not merge before:

- [ ] https://github.com/open-dynamic-robot-initiative/blmc_drivers/pull/4

### CODE REVIEW

File: demos/demo_solo12_calibration.cpp [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    CTRL_C_DETECTED = true;
+    return THREAD_FUNCTION_RETURN_VALUE;
+}  // end control_loop
+
+
+int main(int argc, char** argv)
+{
+    enable_ctrl_c();
+
+    assert(argc == 2 && "Wrong number of argument: ./demo_solo12_calibration network_id");
```
Comment (jviereck) [inline demos/demo_solo12_calibration.cpp]: I always forget: If you compile in optimized mode, are the assertions still executed?

File: demos/demo_solo12_calibration.cpp [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+    std::shared_ptr<Solo12> robot = std::make_shared<Solo12>();
+    robot->initialize(argv[1]);
+
+    ThreadCalibrationData_t thread_data(robot);
+
+    rt_printf("controller is set up \n");
```
Comment (jviereck) [inline demos/demo_solo12_calibration.cpp]: Please fix string to follow proper language gramma/syntax rules (first letter of sentence upper case, add `.` before newline).

File: demos/demo_solo12_calibration.cpp [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    rt_printf("Press enter to launch the calibration \n");
```
Comment (jviereck) [inline demos/demo_solo12_calibration.cpp]: Same here.

File: include/blmc_robots/solo12.hpp [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+#include "blmc_robots/common_header.hpp"
+#include "blmc_robots/blmc_joint_module.hpp"
+
+
+namespace blmc_robots {
+
+/**
+ * @brief Definition and drivers for the Solo12 robot.
+ * 
+ *
```
Comment (jviereck) [inline include/blmc_robots/solo12.hpp]: Remove extra comment line. Same below.

File: include/blmc_robots/solo12.hpp [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+public:
+
+  /**
+   * @brief Solo is the constructor of the class.
+   */
+  Solo12();
+
+  /**
+   * @brief initialize the robot by setting aligning the motors and calibrate the
+   * sensors to 0
```
Comment (jviereck) [inline include/blmc_robots/solo12.hpp]: Comments need grammatical and syntax fixing (colon at end, first letter upper case). Also applies to other comments.

File: include/blmc_robots/solo12.hpp [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+  }
+
+private:
+  /**
+   * Joint properties
+   */
+
+  /** @brief Motors inertia. */
+  Vector12d motor_inertias_;
+  /** @brief DCM motor torque constants. */
```
Comment (jviereck) [inline include/blmc_robots/solo12.hpp]: Want to have give the declarations here some more space and add a new line before the next docstring etc? Applies also to lines below.

File: include/blmc_robots/solo8.hpp [hunk 7]
```diff
@@ -43,18 +43,18 @@ class Solo
...
   /**
    * @brief Calibrate the joints by moving to the next joint index position.
-   * 
+   *
    * @param home_offset_rad This is the angle between the index and the zero
    * pose.
-   * @return true 
-   * @return false 
+   * @return true
+   * @return false
```
Comment (jviereck) [inline include/blmc_robots/solo8.hpp]: Return bool?

File: src/solo12.cpp [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+  // Initialize the communication with the main board.
+  main_board_ptr_ = std::make_shared<MasterBoardInterface>(network_id_);
+  main_board_ptr_->Init();
+
+  // create the SpiBus (blmc_drivers wrapper around the MasterBoardInterface)
+  spi_bus_ = std::make_shared<blmc_drivers::SpiBus>(main_board_ptr_,
+                                                    motor_boards_.size());
+
+  // create the motor board objects:
+  for(size_t mb_id = 0 ; mb_id < motor_boards_.size() ; ++mb_id)
```
Comment (jviereck) [inline src/solo12.cpp]: Please fix formating of the `for` statement. Also below.

Maybe run clang-format in general on this PR?

File: src/solo12.cpp [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+  reverse_polarities_[2] = true;  // FL_KFE
+  reverse_polarities_[3] = false; // FR_HAA
+  reverse_polarities_[4] = false; // FR_HFE
+  reverse_polarities_[5] = false; // FR_KFE
+  reverse_polarities_[6] = true;  // HL_HAA
+  reverse_polarities_[7] = true;  // HL_HFE
+  reverse_polarities_[8] = true;  // HL_KFE
+  reverse_polarities_[9] = false;  // HL_HAA
+  reverse_polarities_[10] = false; // HR_HFE
+  reverse_polarities_[11] = false; // HR_KFE
```
Comment (jviereck) [inline src/solo12.cpp]: Can you assign via `reverse_polarities_ = {true, true, ...}`?

File: src/solo12.cpp [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+  joints_.set_joint_polarities(reverse_polarities_);
+
+  // The the control gains in order to perform the calibration
+  blmc_robots::Vector12d kp, kd;
+  kp.fill(3.0);
+  kd.fill(0.05);
+  joints_.set_position_control_gains(kp, kd);
+
+
+  // Wait until all the motors are ready.
```
Comment (jviereck) [inline src/solo12.cpp]: Extra newline.

File: srcpy/py_solo8.cpp [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+            &Solo8::set_max_joint_torques,
+            py::arg("max_joint_torques"))
+        .def("get_motor_board_errors", &Solo8::get_motor_board_errors)
+        .def("get_motor_board_enabled", &Solo8::get_motor_board_enabled)
+        .def("get_motor_enabled", &Solo8::get_motor_enabled)
+        .def("get_motor_ready", &Solo8::get_motor_ready)
+        .def("get_slider_positions", &Solo8::get_slider_positions)
+        .def("get_joint_positions", &Solo8::get_joint_positions)
+        .def("get_joint_velocities", &Solo8::get_joint_velocities);
+}
```
Comment (jviereck) [inline srcpy/py_solo8.cpp]: Question: Have you tried if these work? I wrote the code but not sure I've tested it.

File: demos/demo_solo12_calibration.cpp (hunk 1 above)
Comment (MaximilienNaveau) [reply demos/demo_solo12_calibration.cpp]: no, you prefer having an exception?

File: src/solo12.cpp (hunk 8 above)
Comment (MaximilienNaveau) [reply src/solo12.cpp]: I will progress on the formatting on a different PR. I would need to check the robot behaviors after the general formatting.

File: srcpy/py_solo8.cpp (hunk 11 above)
Comment (MaximilienNaveau) [reply srcpy/py_solo8.cpp]: I did not tested it yet. To be checked for another PR?

File: src/programs/solo12_hardware_calibration.cpp [hunk 12]
```diff
@@ -36,7 +36,10 @@ static THREAD_FUNCTION_RETURN_TYPE control_loop(void* robot_void_ptr)
 
 int main(int argc, char** argv)
 {
-    assert(argc == 2 && "Wrong number of argument: ./demo_solo12_calibration network_id");
+    if(argc == 2)
```
Comment (jviereck) [inline src/programs/solo12_hardware_calibration.cpp]: Are you sure this is correct? Shouldn't it be `argc != 2`?

**Pull request merged**